pkg/fqdn/name_manager.go

```diff
@@ -179,20 +198,31 @@ func (n *NameManager) UpdateGenerateDNS(ctx context.Context, lookupTime time.Tim
 		}).Debug("Updated FQDN with new IPs")
 	}
 
-	namesMissingIPs, selectorIPMapping := n.GenerateSelectorUpdates(fqdnSelectorsToUpdate)
+	namesMissingIPs, selectorIPMapping := n.generateSelectorUpdates(fqdnSelectorsToUpdate)
 	if len(namesMissingIPs) != 0 {
 		log.WithField(logfields.DNSName, namesMissingIPs).
 			Debug("No IPs to insert when generating DNS name selected by ToFQDN rule")
 	}
 
-	return n.config.UpdateSelectors(ctx, selectorIPMapping, namesMissingIPs)
+	errCh = make(chan error, 1)
+	n.updateQueue <- func() {
```

Comment:
This will block if the queue is full; should we attempt to send, and break out / immediately return an error if we somehow get into that adverse condition?

At least that way we should be able to present a log message to the user that states that these DNS IP mappings are not being propagated, which would eventually trickle through to us to tell us we need to iterate further on this (rather than potentially getting deadlocked like the current code does)

If we do this, we may also consider trying to roll back the changes that are unsuccessful, with the idea that if the endpoint retries to learn this DNS name then we'll learn it next time..?

Reply:
I managed the logging but rolling back is more difficult. Once the cache is updated undoing that isn't directly possible (we could do it by making a scratch cache, doing all the updates there, and only committing if this succeeds. This is a lot more expensive, though, so it might not be worth it. I've upped the queue to 65k entries. I imagine that we would be hard-deadlocked for some other reason to reach that state